### ISSUE TITLE
Speed up via xformers

### ISSUE THREAD

**Issue by mortonjt** [agemagician/ProtTrans#120]
Just in case you weren't familiar with this, there is an xformers library that can allow for a >4x speed up on all transformer operations
https://github.com/facebookresearch/xformers

Could be low hanging fruit to speed up the operations in this library :)

**Comment by mheinzinger** [comment]
Hi Jamie,

thanks for reaching out! - I wanted to try this before answering but obviously it took me way too long. I already gave it a shot few weeks ago but failed to reach some significant speed-up but maybe I did something wrong.

Have you made positive experience with this using some protein language models?

**Comment by mortonjt** [comment]
Sorry to hear about that.  I haven't tried this out yet for protein LLMs (only tested it out on stable-diffusion), but it is on my radar.  Hoping that it could be useful for inference and speed up the embedding calculations (which we're noticing is a bottleneck for protein annotation)

**Comment by mheinzinger** [comment]
Hm, how many proteins are you trying to label? - From my experience ProtT5-XL-U50 encoder-only in half-precision using batching as described [here](https://github.com/agemagician/ProtTrans/blob/master/Embedding/prott5_embedder.py) reaches around 0.1s/protein on average for the 20k proteins human (so around 30m for human).

**Comment by mheinzinger** [comment]
I had a brief look and I stopped once I hit the following error: `AttributeError: 'FeatureExtractionPipeline' object has no attribute 'enable_xformers_memory_efficient_attention'` (tried to extract embeddings from the ProtT5-XL-U50-fp16 model from my link in the post above). 
So not sure whether it is as easily plug-n-play as I had hoped. In case you find some example/tutorial that shows how this should be done for plain Transformers (no diffusion etc), pls send by and I can give it a try. So far, I only found tutorials on how to use this on diffusion models in huggingface (but most likely I just missed the right source)

**Comment by mortonjt** [comment]
Regarding examples, I first saw xformers being used in
https://github.com/Stability-AI/stablediffusion — so yes I only saw this
used in diffusion models

We were trying to embed all of uniref at one point, but had to resort to
just a subset. Were trying to embed proteins in microbial metagenomes, and
those reference databases are often >50M proteins

**Comment by mheinzinger** [comment]
Yeah, I see your point. We also ran UniRef50 at one point but only to make predictions, not for embedding extraction (esp. as storing those embeddings becomes expensive quickly).
Only things I can recommend (probably obvious but still):
- sort & process sequences by length to avoid padding.
- batch-processing & half-precision to max-out batch sizes
- If you write chunks of proteins (after sorting by length), you can parallelize embedding extraction on multiple GPUs
- Maybe check [TensorRT T5-example](https://github.com/NVIDIA/TensorRT/blob/5f422623e7f5bdc593b781695cbddda99124c9b8/demo/HuggingFace/notebooks/t5.ipynb) (no experience, though)
